From 842a74d19db603197002d67704c9c4dcb4ff1fcb Mon Sep 17 00:00:00 2001
From: Chen Li1 <li1.chen@intel.com>
Date: Tue, 6 Mar 2018 14:56:15 +0800
Subject: [PATCH] Make webrtcConnection::close async to avoid
 deliver[Video/Audio]Data after destructing. May remove this patch after next
 licode upgrading.

---
 erizo/src/erizo/WebRtcConnection.cpp | 11 ++++++++---
 erizo/src/erizo/WebRtcConnection.h   |  1 +
 2 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/erizo/src/erizo/WebRtcConnection.cpp b/erizo/src/erizo/WebRtcConnection.cpp
index b793cb3..2520317 100644
--- a/erizo/src/erizo/WebRtcConnection.cpp
+++ b/erizo/src/erizo/WebRtcConnection.cpp
@@ -76,13 +76,18 @@ WebRtcConnection::WebRtcConnection(std::shared_ptr<Worker> worker, std::shared_p
 
 WebRtcConnection::~WebRtcConnection() {
   ELOG_DEBUG("%s message:Destructor called", toLog());
-  if (sending_) {
-    close();
-  }
   ELOG_DEBUG("%s message: Destructor ended", toLog());
 }
 
 void WebRtcConnection::close() {
+  ELOG_DEBUG("%s message: Async close called", toLog());
+  auto conn_ptr = shared_from_this();
+  worker_->task([conn_ptr]{
+    conn_ptr->syncClose();
+  });
+}
+
+void WebRtcConnection::syncClose() {
   ELOG_DEBUG("%s message:Close called", toLog());
   if (!sending_) {
     return;
diff --git a/erizo/src/erizo/WebRtcConnection.h b/erizo/src/erizo/WebRtcConnection.h
index b50eda4..5793877 100644
--- a/erizo/src/erizo/WebRtcConnection.h
+++ b/erizo/src/erizo/WebRtcConnection.h
@@ -84,6 +84,7 @@ class WebRtcConnection: public MediaSink, public MediaSource, public FeedbackSin
    */
   bool init();
   void close() override;
+  void syncClose();
   /**
    * Sets the SDP of the remote peer.
    * @param sdp The SDP.
-- 
1.9.1

